heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,231 @@
+    return 'error';
+  }
+  return 'warning';
+}
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+interface ISarifRun {
+  tool: {
+    driver: {
+      name: string;
+      informationUri: string;
+      version?: string;
+      rules: any[];
+    };
+  };
+  artifacts?: any[];
+  results?: any[];
+  invocations?: {
+    toolConfigurationNotifications: any[];
+    executionSuccessful: boolean;
+  }[];
+}
+
+// Interface for the SARIF log structure
+interface ISarifLog {
+  version: string;
+  $schema: string;
+  runs: ISarifRun[];
+}
+
+export interface ISerifFormatterOptions {
+  ignoreSuppressed: boolean;
+  eslintVersion: string;
+}
+
+interface IExtendedLintResult extends TEslint.ESLint.LintResult {
+  suppressedMessages?: TEslint.ESLint.LintResult['messages'];
+}
+
+// Main function
+export function formatAsSARIF(results: IExtendedLintResult[], options: ISerifFormatterOptions): ISarifLog {
+  const sarifRun: ISarifRun = {
+    tool: {
+      driver: {
+        name: 'ESLint',
+        informationUri: 'https://eslint.org',
+        rules: []
+      }
+    }
+  };
+  const sarifLog: ISarifLog = {
+    version: '2.1.0',
+    $schema: 'http://json.schemastore.org/sarif-2.1.0-rtm.5',
+    runs: [sarifRun]
+  };
+
+  const { ignoreSuppressed, eslintVersion } = options;
+
+  if (typeof eslintVersion !== 'undefined') {
+    sarifRun.tool.driver.version = eslintVersion;
+  }
+
+  const sarifFiles: Map<string, ISarifFile> = new Map();
+  const sarifResults: any[] = [];
+
+  const internalErrorId = 'ESL0999';
+  const toolConfigurationNotifications: any[] = [];
+  let executionSuccessful = true;
+
+  for (const result of results) {
+    const { filePath } = result;
+    let sarifFile: ISarifFIle | undefined = sarifFiles.get(filePath);
+    if (sarifFile === undefined) {
+      const artifactIndex: number = sarifFiles.size;
+      const fileUrl: string = url.pathToFileURL(filePath).toString();
+
+      sarifFile = {
+        location: {
+          uri: fileUrl
+        }
+      };
+      sarifFiles.set(filePath, sarifFile);
+
+      const containsSuppressedMessages = result.suppressedMessages && result.suppressedMessages.length > 0;
+      const messages =
+        containsSuppressedMessages && !ignoreSuppressed
+          ? [...result.messages, ...result.suppressedMessages]
+          : result.messages;
+
+      if (messages.length > 0) {
+        for (const message of messages) {
+          const sarifRepresentation: any = {
+            level: getResultLevel(message),
+            message: {
+              text: message.message
+            },
+            locations: [
+              {
+                physicalLocation: {
+                  artifactLocation: {
+                    uri: fileUrl,
+                    index: artifactIndex
+                  }
+                }
```

Comment:
Put the `physicalLocation` property in a local variable and refer to it later in all the places that are accessing `sarifRepresentation.locations[0].physicalLocation`.